sky/jobs/controller.py

```diff
@@ -577,199 +779,436 @@ def _update_failed_task_state(
-        return
-    assert user_signal == managed_job_utils.UserSignal.CANCEL, (
-        f'Only cancel signal is supported, but {user_signal} got.')
-    raise exceptions.ManagedJobUserCancelledError(
-        f'User sent {user_signal.value} signal.')
-
-
-def _cleanup(job_id: int, dag_yaml: str, pool: Optional[str]):
-    """Clean up the cluster(s) and storages.
-
-    (1) Clean up the succeeded task(s)' ephemeral storage. The storage has
-        to be cleaned up after the whole job is finished, as the tasks
-        may share the same storage.
-    (2) Clean up the cluster(s) that are not cleaned up yet, which can happen
-        when the task failed or cancelled. At most one cluster should be left
-        when reaching here, as we currently only support chain DAGs, and only
-        task is executed at a time.
-    """
-    # Cleanup the HA recovery script first as it is possible that some error
-    # was raised when we construct the task object (e.g.,
-    # sky.exceptions.ResourcesUnavailableError).
-    managed_job_state.remove_ha_recovery_script(job_id)
-    dag, _ = _get_dag_and_name(dag_yaml)
-    for task in dag.tasks:
-        assert task.name is not None, task
-        if pool is None:
-            cluster_name = managed_job_utils.generate_managed_job_cluster_name(
-                task.name, job_id)
-            managed_job_utils.terminate_cluster(cluster_name)
-        else:
-            cluster_name, job_id_on_pool_cluster = (
-                managed_job_state.get_pool_submit_info(job_id))
-            if cluster_name is not None:
-                if job_id_on_pool_cluster is not None:
-                    core.cancel(cluster_name=cluster_name,
-                                job_ids=[job_id_on_pool_cluster],
-                                _try_cancel_if_cluster_is_init=True)
-
-        # Clean up Storages with persistent=False.
-        # TODO(zhwu): this assumes the specific backend.
-        backend = cloud_vm_ray_backend.CloudVmRayBackend()
-        # Need to re-construct storage object in the controller process
-        # because when SkyPilot API server machine sends the yaml config to the
-        # controller machine, only storage metadata is sent, not the storage
-        # object itself.
-        for storage in task.storage_mounts.values():
-            storage.construct()
-        backend.teardown_ephemeral_storage(task)
-
-        # Clean up any files mounted from the local disk, such as two-hop file
-        # mounts.
-        for file_mount in (task.file_mounts or {}).values():
+class Controller:
+    """Controller for managing jobs."""
+
+    def __init__(self):
+        # Global state for active jobs
+        self.job_tasks: Dict[int, asyncio.Task] = {}
+        self.starting: Set[int] = set()
+
+        # Lock for synchronizing access to global state dictionary
+        self._job_tasks_lock = asyncio.Lock()
+        self._starting_signal = asyncio.Condition(lock=self._job_tasks_lock)
```

Comment:
Add a comment about when we should signal (e.g. whenever a new job may be able to start or whatever)